### ISSUE TITLE
Getting issue when i import in my project ERROR: ENOENT: no such file or directory, open '//../base/shared/util.js'

### ISSUE THREAD

**Issue by astha0211** [modesty/pdf2json#161]
ERROR: ENOENT: no such file or directory, open '//../base/shared/util.js'
chrome
    at Object.fs.openSync (fs.js:663:18)
    at Object.fs.readFileSync (fs.js:568:33)
    at pdfjsFiles.forEach (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:29632:66)
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:29632:13)
    at Object.nameStartChar (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:30069:30)
    at __webpack_require_ (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:20:30)
    at Object.<anonymous> (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:23766:10)
    at _webpack_require_ (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:20:30)
    at Object.defineProperty.value (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:23580:17)
    at _webpack_require_ (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:20:30)
    at Object.<anonymous> (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:21353:1)
    at _webpack_require_ (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:20:30)
    at Object.<anonymous> (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:21341:18)
    at _webpack_require_ (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:20:30)
    at exports.__esModule (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:63:18)
    at Object.<anonymous> (/Users/sshekhar/dev/transactions-sfc-ui-v2/build/js/testSuites.js:66:10)
    at Module._compile (module.js:660:30)
    at Object.Module._extensions..js (module.js:671:10)
    at Module.load (module.js:573:32)
    at tryModuleLoad (module.js:513:12)
    at Function.Module._load (module.js:505:3)

****************************
In the following file /lib/pdf.js  

`let _basePath = __dirname + "/../base/";
let _fileContent = '';

_pdfjsFiles.forEach( (fieldName, idx, arr) => _fileContent += fs.readFileSync(_basePath + fieldName, 'utf8') );`

In My project i am doing something
import PDFParser from "pdf2json";

**Comment by cclark25** [comment]
I have the same problem. I am trying to use pdf2json in an AWS lambda using serverless. __dirname is a node internal variable that is meant to get the directory that the active file is in, which is "/lib" in this case. In my case, the __dirname variable seems to be altered by the Webpack JS bundler. Apparently the default __dirname for webpack is set to '/'.
In the webpack.config.js file, I tried to fix the problem by adding the line:

`target: 'node',
    node: {
        __dirname: false,
    },`

But that only changed the __dirname string to the directory where the file is required from rather than the actual folder where the pdf.js file is located, causing the same error because the ../base directory does not exist there.
I also tried setting __dirname to true instead of false in the webpack.config.js file, which results in the __dirname string to be "node_modules/pdf2json/lib". Since this is still not an absolute file path, the error still persists.

**Comment by ddennis** [comment]
I have the same problem - when trying to run the function on "netlify functions". 
It would be greatly appreciated if any one could provide info on how to fix or work around it,

```
{
"errorType": "Error",
"errorMessage": "ENOENT: no such file or directory, open '//../base/shared/util.js'",
"trace": [
"Error: ENOENT: no such file or directory, open '//../base/shared/util.js'",
"    at Object.openSync (fs.js:458:3)",
"    at Object.readFileSync (fs.js:360:35)",
"    at /var/task/src/server.js:328:42598",
"    at Array.forEach (<anonymous>)",
"    at Object.<anonymous> (/var/task/src/server.js:328:42564)",
"    at Object.<anonymous> (/var/task/src/server.js:328:49100)",
"    at e (/var/task/src/server.js:1:158)",
"    at Object.<anonymous> (/var/task/src/server.js:328:14447)",
"    at e (/var/task/src/server.js:1:158)",
"    at Object.<anonymous> (/var/task/src/server.js:328:13643)"
]
}
```


Kind regards
Dennis

**Comment by rlecaro2** [comment]
### Context
This is the "offending" code. I guess this is done to share context between several javascript files. Patching this is not trivial, and several files work under this _global scope_ and shared variables assumption. 

```javascript
// lib/pdf.js
let PDFJS = {};
let globalScope = {console: console};

let _basePath = __dirname + "/../base/";
let _fileContent = '';

_pdfjsFiles.forEach( (fieldName, idx, arr) => _fileContent += fs.readFileSync(_basePath + fieldName, 'utf8') );

eval(_fileContent);
```

Webpack has a hard time with dynamic imports, and should never waste time supporting this use case (that is way crazier). A quick fix for anyone interested is just skipping this package in webpack and copy it over to wherever your bundle is.

### A webpack workaround
I used [copy-webpack-plugin](https://webpack.js.org/plugins/copy-webpack-plugin/)
and added `pdf2json` as an external dependency
```javascript
// webpack.config.js
const CopyPlugin = require('copy-webpack-plugin');

const dontBundlePdf2Json = new CopyPlugin({
  patterns: ['node_modules/pdf2json/**/*'],
});

const externals = ['pdf2json'];

module.exports = {
  /* all the other configuration */
  externals,
  plugins: [dontBundlePdf2Json],
}
```

This will result in:
```
project/
├── dist/
│   ├── node_modules/
│   │   └── pdf2json/
│   │       └── ... all the library's files
│   └── bundle.js
└─── src/ 
```

And the node module resolution works as expected.

**Comment by xMGZx** [comment]
Bruh why you use wacky module resolution _and_ eval

**Comment by JinhooBong** [comment]
Has the copy-webpack-plugin solution worked for anyone? It has not resolved the issue for me.

**Comment by isitgeorge** [comment]
I'm having this same issue when packaging an Electron app - the webpack copy plugin workaround hasn't worked but I'm still tinkering.

```
Uncaught Exception:
Error: ENOENT, dist/base/shared/util.js not found in <project>/release/build/mac-arm64/ElectronReact.app/Contents/Resources/app.asar
at createError (node:electron/js2c/asar_bundle:2:1312)
```

Are there any other trivial workarounds?

**Comment by modesty** [comment]
A fix plus appropriate tests are in [PR #338](https://github.com/modesty/pdf2json/pull/338), please review and test it out

**Comment by modesty** [comment]
fixed in 3.1.2

**Comment by isitgeorge** [comment]
@modesty Thank you for fixing!

I've only just had some time to test and it seems to work without issue when packaged as part of an Electron app.
